sdk-core/src/main/java/care/data4life/sdk/migration/RecordCompatibilityService.kt

```diff
@@ -19,138 +19,101 @@ package care.data4life.sdk.migration
+            tagGroupKey
         )
     }
 
-    private fun needsDoubleCall(
-        encodedAndEncryptedTags: List<String>,
-        encryptedTags: List<String>
-    ): Boolean = encodedAndEncryptedTags.sorted() != encryptedTags.sorted()
+    private fun addTags(
+        tagGroupKey: String,
+        tagOrGroup: Triple<String, String, String>,
+        tagEncryptionKey: GCKey,
+        builder: NetworkingContract.SearchTagsBuilder
+    ) {
+        builder.addOrTuple(
+            encryptTags(
+                tagEncryptionKey,
+                tagGroupKey,
+                tagOrGroup.toList()
+            )
+        )
+    }
 
-    override fun searchRecords(
-        alias: String,
-        userId: String,
-        startDate: String?,
-        endDate: String?,
-        pageSize: Int,
-        offSet: Int,
+    private fun mapTags(
         tags: Tags,
-        annotations: Annotations
-    ): Observable<List<EncryptedRecord>> {
-        val (encodedAndEncryptedTags, encryptedTags) = encrypt(tags, annotations)
-        return if (needsDoubleCall(encodedAndEncryptedTags, encryptedTags)) {
-            Observable.zip(
-                apiService.searchRecords(
-                    alias,
-                    userId,
-                    startDate,
-                    endDate,
-                    pageSize,
-                    offSet,
-                    encodedAndEncryptedTags.joinToString(",")
-                ),
-                apiService.searchRecords(
-                    alias,
-                    userId,
-                    startDate,
-                    endDate,
-                    pageSize,
-                    offSet,
-                    encryptedTags.joinToString(",")
-                ),
-                BiFunction<List<EncryptedRecord>, List<EncryptedRecord>, List<EncryptedRecord>> { records1, records2 ->
-                    mutableListOf<EncryptedRecord>().also {
-                        it.addAll(records1)
-                        it.addAll(records2)
-                    }
-                }
+        tagEncryptionKey: GCKey,
+        builder: NetworkingContract.SearchTagsBuilder
+    ) {
+        tags.map { tagGroup ->
+            Pair(
+                tagGroup.key + DELIMITER,
+                compatibilityEncoder.encode(tagGroup.value)
+            )
+        }.map { encodedTagGroup ->
+            addTags(
+                encodedTagGroup.first,
+                encodedTagGroup.second,
+                tagEncryptionKey,
+                builder
+            )
+        }
+    }
+
+    private fun mapAnnotations(
+        annotations: Annotations,
+        tagEncryptionKey: GCKey,
+        builder: NetworkingContract.SearchTagsBuilder
+    ) {
+        annotations.map { annotation ->
+            Pair(
+                ANNOTATION_KEY + DELIMITER,
+                compatibilityEncoder.encode(annotation).copy(
+                    second = compatibilityEncoder.normalize(annotation)
```

Comment:
why is the second one normalized twice? doesn't encode normalize the second parameter here already?

Reply:
It makes a copy of the existing Triple and replaces the KMP legacy annotations, but they [should not normalized at all](https://github.com/d4l-data4life/hc-sdk-kmp/pull/49/files).

Reply:
fixed.